### PR TITLE
Add a CMAP test that verifies the background thread hands over connections to threads doing checkout

### PR DESCRIPTION
**Pull request by stIncMale**
Added @ShaneHarvey as a reviewer because introducing this test was his idea.

DRIVERS-1948

### CODE REVIEW

File: source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    failCommands: ["isMaster","hello"]
+    closeConnection: false
+    blockConnection: true
+    blockTimeMS: 500
+poolOptions:
+  # allows both thread1 and the background thread to start opening connections concurrently
+  minPoolSize: 2
+  # gives opportunity for the checkout in thread2 to open a new connection, which it must not do nonetheless
+  maxPoolSize: 3
+  waitQueueTimeoutMS: 5000
```
Comment (ShaneHarvey) [inline source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml:21]: Do we need backgroundThreadIntervalMS here?

**Comment by stIncMale** [comment]
> This test fails in Java before the fix in JAVA-4316?

Yes, it fails without the changes in https://github.com/mongodb/mongo-java-driver/pull/799, which encompass both JAVA-4316 and JAVA-4316: _"expected:<class `com.mongodb.event.ConnectionCheckedOutEvent`> but was:<class `com.mongodb.event.ConnectionCreatedEvent`>"_.

File: source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml (hunk 1 above)
Comment (stIncMale) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml:21]: No, all we need is for the BG thread to start a single [run](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#background-thread), which it does without artificial delays, and then we do not care when the next run starts.

File: source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
+version: 1
+style: integration
+description: threads blocked by maxConnecting check out minSize connections
```
[the remark this answers is not in the document]
Comment (stIncMale) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml]: Done

File: source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml (hunk 1 above)
Comment (stIncMale) [reply source/connection-monitoring-and-pooling/tests/pool-checkout-minSize-connection-maxConnecting.yml]: It is not explicitly specified that the background thread should be started ASAP ("immediately" is not an appropriate term here), but the specification of `minPoolSize` (which is unachievable, and, therefore, should have been formulated differently) implies this: [_"The minimum number of Connections that MUST exist at any moment in a single connection pool."_](https://github.com/mongodb/specifications/blob/master/source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst#id1)

File: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst [hunk 3]
```diff
@@ -523,6 +523,16 @@ occur only while the pool is "ready". If the pool implements a background
 thread, it can be used for this. If the pool does not implement a background
 thread, the checkOut method is responsible for ensuring this requirement is met.
 
+When populating the Pool, pendingConnectionCount has to be decremented after
+establishing a `Connection`_ similarly to how it is done in
```
Comment (ShaneHarvey) [inline source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst:527]: In my mind "after establishing a `Connection`_" does not convey the nuance required here. I think we need to say that pendingConnectionCount must not be decremented until after the established Connection is actually added to the pool.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst (hunk 3 above)
Comment (stIncMale) [reply source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst:527]: The first sentence is not there to specify a requirement (even though it must be so), but to introduce the notion of a specific signal, for which an ordering constraint is then specified. I updated the ordering constraint such that it refers to the "Marking a Connection as Available" action, which is the wording the specification uses to say "add connection to the pool".

File: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst [hunk 4]
```diff
@@ -523,6 +523,17 @@ occur only while the pool is "ready". If the pool implements a background
...
+`Checking Out a Connection <#checking-out-a-connection>`_ to signal that
+another `Connection`_ is allowed to be established. Such a signal must become
```
Comment (ShaneHarvey) [inline source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst:529]: must -> MUST

File: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst (hunk 3 above)
Comment (ShaneHarvey) [reply source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst:527]: Ah thanks I get it now.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]
One minor nit otherwise LGTM

**Review by ShaneHarvey — APPROVED** [review]
LGTM

**Review by BorisDog — APPROVED** [review]
LGTM

**Pull request merged**